compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -613,6 +619,43 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             }),
         }
     }
+
+    /// Detect the inference regression on crate `time` <= 0.3.35 and emit a more targeted error.
+    /// <https://github.com/rust-lang/rust/issues/127343>
+    // FIXME: we should figure out a more generic version of doing this, ideally in cargo itself.
+    fn detect_old_time_crate_version(
+        &self,
+        span: Option<Span>,
+        kind: &InferSourceKind<'_>,
+        // We will clear the non-actionable suggestion from the error to reduce noise.
+        infer_subdiags: &mut Vec<SourceKindSubdiag<'_>>,
+    ) -> Option<()> {
+        if self.infcx.tcx.crate_name(LOCAL_CRATE) != sym::time {
+            // Only relevant when building the `time` crate.
+            return None;
+        }
+        let Some(span) = span else {
+            return None;
+        };
+        let InferSourceKind::LetBinding { pattern_name, .. } = kind else {
+            return None;
+        };
+        let Some(name) = pattern_name else {
+            return None;
+        };
+        if name.as_str() != "items" {
+            return None;
+        }
+        let FileName::Real(file) = self.infcx.tcx.sess.source_map().span_to_filename(span) else {
+            return None;
+        };
+        let path = file.local_path_if_available().to_string_lossy();
+        if path.contains("format_description") && path.contains("parse") {
+            infer_subdiags.clear();
+            return Some(());
+        }
+        None
```

Comment:
This could be made much nicer (IMHO) with let chains:

```suggestion
        if self.infcx.tcx.crate_name(LOCAL_CRATE) == sym::time
            && let Some(span) = span
            && let InferSourceKind::LetBinding { pattern_name: Some(name), .. } = kind
            && name == "items"
            && let FileName::Real(file) = self.infcx.tcx.sess.source_map().span_to_filename(span)
            && {
                let path = file.local_path_if_available().to_string_lossy();
                path.contains("format_description") && path.contains("parse")
            }
        {
            infer_subdiags.clear();
            Some(())
        } else {
            None
        }
```

Reply:
Discussion (for @estebank): I think this does make it easier to follow what conditions are needed to trigger the diagnostic, but I'm not going to block on that. At your discretion, whatever is easier to read for you.